File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res =0;

    int maxAncestorDiff(TreeNode* root) {
        int mx = INT_MIN;
        int mn = INT_MAX;
        pre(root , mx , mn);
        return res;
    }

    void pre(TreeNode* node , int mx , int mn){
        if(!node){
            return ;
        }
        if(mx!=INT_MIN) res = max(res , abs(mx - node->val));
        if(mn!=INT_MAX) res = max(res , abs(node->val - mn));

        mx = max(mx , node->val);
        mn = min(mn , node->val);
        pre(node->left , mx , mn);
        pre(node->right , mx , mn);
    }
};

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        self.res = 0
        self.dfs(root, root.val, root.val)
        return self.res

    def dfs(self, node, maxSeen, minSeen):
        self.res = max(abs(node.val-maxSeen), self.res, abs(node.val-minSeen))
        maxSeen = max(maxSeen, node.val)
        minSeen = min(minSeen, node.val)

        if node.left:
            self.dfs(node.left, maxSeen, minSeen)
        if node.right:
            self.dfs(node.right, maxSeen, minSeen)
        